form/form_test.go

```diff
@@ -0,0 +1,257 @@
+package form
+
+import (
+	"testing"
+
+	assert "github.com/stretchr/testify/require"
+)
+
+type testStruct struct {
+	Array    [3]string  `form:"array"`
+	ArrayPtr *[3]string `form:"array_ptr"`
+
+	ArrayIndexed    [3]string  `form:"array_indexed,indexed"`
+	ArrayIndexedPtr *[3]string `form:"array_indexed_ptr,indexed"`
+
+	Bool    bool  `form:"bool"`
+	BoolPtr *bool `form:"bool_ptr"`
+
+	Emptied bool `form:"emptied,empty"`
+
+	Float32    float32  `form:"float32"`
+	Float32Ptr *float32 `form:"float32_ptr"`
+
+	Float64    float64  `form:"float64"`
+	Float64Ptr *float64 `form:"float64_ptr"`
+
+	Int      int    `form:"int"`
+	IntPtr   *int   `form:"int_ptr"`
+	Int8     int8   `form:"int8"`
+	Int8Ptr  *int8  `form:"int8_ptr"`
+	Int16    int16  `form:"int16"`
+	Int16Ptr *int16 `form:"int16_ptr"`
+	Int32    int32  `form:"int32"`
+	Int32Ptr *int32 `form:"int32_ptr"`
+	Int64    int64  `form:"int64"`
+	Int64Ptr *int64 `form:"int64_ptr"`
+
+	Inverted bool `form:"inverted,invert"`
+
+	Map map[string]interface{} `form:"map"`
+
+	Slice    []string  `form:"slice"`
+	SlicePtr *[]string `form:"slice_ptr"`
+
+	String    string  `form:"string"`
+	StringPtr *string `form:"string_ptr"`
+
+	SliceIndexed    []string  `form:"slice_indexed,indexed"`
+	SliceIndexedPtr *[]string `form:"slice_indexed_ptr,indexed"`
+
+	SubStruct    testSubStruct  `form:"substruct"`
+	SubStructPtr *testSubStruct `form:"substruct_ptr"`
+
+	Uuint      uint    `form:"uint"`
+	UuintPtr   *uint   `form:"uint_ptr"`
+	Uuint8     uint8   `form:"uint8"`
+	Uuint8Ptr  *uint8  `form:"uint8_ptr"`
+	Uuint16    uint16  `form:"uint16"`
+	Uuint16Ptr *uint16 `form:"uint16_ptr"`
+	Uuint32    uint32  `form:"uint32"`
+	Uuint32Ptr *uint32 `form:"uint32_ptr"`
+	Uuint64    uint64  `form:"uint64"`
+	Uuint64Ptr *uint64 `form:"uint64_ptr"`
+
+	Zeroed bool `form:"zeroed,zero"`
+}
+
+type testSubStruct struct {
+	SubSubStruct testSubSubStruct `form:"subsubstruct"`
```

Comment:
Basically anything that's not a leaf primitive type with a `form` tag will get ignored.

So while a `string` under `testSubSubStruct` gets serialized as expected here, these would be ignored:

1. A `struct` without a `form` tag.
2. A `string` (or other primitive) with a `form` tag that's nested under a struct _without_ a `form` tag (we need a valid parent name to assemble a form key hierarchy like `param[struct][substruct][foo]...`).